### PR TITLE
Rails blank

### PR DESCRIPTION
**Pull request by rrosenblum**
Add 2 new cops `Rails/Blank` and `Rails/Present`. Both of these cops look for common checks that can be converted to one or another, inverted usages, and usage with unless.

### CODE REVIEW

File: config/enabled.yml [hunk 1]
```diff
@@ -1517,6 +1517,16 @@ Rails/ActiveSupportAliases:
                   `Array#append`, `Array#prepend`.
   Enabled: true
 
+Rails/Blank:
+  Description: 'Enforce using `blank?` and `present?`.'
+  Enabled: true
+  NilOrEmpty: true
+  # Convert checks for `nil` or `empty?` to `blank?`
```
Comment (bbatsov) [inline config/enabled.yml]: I think those comments should be above the option they describe.

File: lib/rubocop/cop/rails/blank.rb [hunk 2]
```diff
@@ -0,0 +1,109 @@
+# frozen_string_literal: true
+
+module RuboCop
+  module Cop
+    module Rails
+      # This cops checks for code that can be changed to `blank?`.
+      # Settings:
+      #   NilOrEmpty: true
+      #     Convert checks for `nil` or `empty?` to `blank?`
```
Comment (bbatsov) [inline lib/rubocop/cop/rails/blank.rb]: Same here.

File: lib/rubocop/cop/rails/blank.rb [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+      #   NotPresent: true
+      #     Convert usages of not `present?` to `blank?`
+      #   UnlessPresent: true
+      #     Convert usages of `unless` `present?` to `blank?`
+      #
+      # @example
+      #   # bad
```
Comment (bbatsov) [inline lib/rubocop/cop/rails/blank.rb]: I'd have good and bad section for each option here. That would make it easier to understand what they actually do.

File: lib/rubocop/cop/rails/blank.rb [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+      #   foo.empty?
+      #   foo.blank?
+      #   foo.present?
+      #   !foo.blank?
+      #   something unless foo.blank?
+      #   somethig if foo.blank?
+      #   something if foo.present?
+      class Blank < Cop
+        MSG_NIL_OR_EMPTY = 'Use `blank?` instead of `nil?` or `empty?`.'.freeze
+        MSG_NOT_PRESENT = 'Use `blank?` instead of not `present?`.'.freeze
```
Comment (bbatsov) [inline lib/rubocop/cop/rails/blank.rb]: `!present?` perhaps?

File: lib/rubocop/cop/rails/blank.rb [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+        MSG_UNLESS_PRESENT =
+          'Use `if` `blank?` instead of `unless` `present?`.'.freeze
```
Comment (bbatsov) [inline lib/rubocop/cop/rails/blank.rb]: `if blank?` `unless present?`

I'd actually include the actual receivers in all those message - that would make them way clearer.

File: spec/rubocop/cop/rails/blank_spec.rb [hunk 6]
```diff
@@ -0,0 +1,264 @@
...
+  it_behaves_like :all_variables, 'foo.bar'
+  it_behaves_like :all_variables, 'FOO'
+  it_behaves_like :all_variables, 'Foo'
+  it_behaves_like :all_variables, '@foo'
+  it_behaves_like :all_variables, '$foo'
+  it_behaves_like :all_variables, '@@foo'
+  it_behaves_like :all_variables, 'foo[bar]'
+  it_behaves_like :all_variables, 'Foo::Bar'
+
+  context 'present?' do
```
Comment (bbatsov) [inline spec/rubocop/cop/rails/blank_spec.rb]: Shouldn't the top-level contexts reflect the configuration options? I find the current structure of the tests a bit random and unclear.

**Comment by bbatsov** [comment]
All the comments I had for `Rails/Blank` apply to the other cop as well.

**Comment by rrosenblum** [comment]
I think that I have addressed all of the comments. Please let me know how the code looks now.

File: lib/rubocop/cop/rails/present.rb [hunk 7]
```diff
@@ -0,0 +1,137 @@
...
+      # Settings:
+      #   NotNilAndNotEmpty: Convert checks for not `nil` and `not empty?`
+      #                      to `present?`
+      #   NotBlank: Convert usages of not `blank?` to `present?`
+      #   UnlessBlank: Convert usages of `unless` `blank?` to `if` `present?`
+      #
+      # @example
+      #   # NotNilAndNotEmpty: true
+      #     # bad
+      #       !foo.nil? && !foo.empty?
```
Comment (bbatsov) [inline lib/rubocop/cop/rails/present.rb]: I'd remove the extra indentation under `bad`/`good`, as it's not something we typically do.

**Comment by bbatsov** [comment]
Apart from one small remark the code looks good. One question - why are the two cops so granular? Seems odd you'd like to use `blank?` or `present?` in just a subset of the places where you'd be able to do so. I didn't pay much attention to this last time around, but now it got me curious.

**Comment by bbatsov** [comment]
@rrosenblum Ping re my question :-)

**Comment by rrosenblum** [comment]
> why are the two cops so granular? 

I assume this is regarding all the configuration options. I originally wasn't going to add them. I know there have been several occasions where after releasing a new cop, we get requests for other styles or configurations. I figured these were simple enough configurations that if someone doesn't like an aspect of the cop, they can easily disable it. 

Also, I had considered adding extra checks, but worried about hitting some edge cases where `blank?` or `present?` might not work as a perfect substitution. Some situations that I think would be safe are `nil? || size == 0`.

**Comment by bbatsov** [comment]
> I assume this is regarding all the configuration options. I originally wasn't going to add them. I know there have been several occasions where after releasing a new cop, we get requests for other styles or configurations. I figured these were simple enough configurations that if someone doesn't like an aspect of the cop, they can easily disable it.

I just don't like overdoing things in advance - it's just as likely that no one would need those configuration options. :-) Anyways, you did the work, so I'll merge this as is, but in general I think we should start with cops that are as simple as possible and add extra configuration only when this is justified.

**Pull request merged**